File: api/companion.py
from http.server import BaseHTTPRequestHandler
import json
import os
import urllib.request
import urllib.error

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        # Enable CORS
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

        # Read request body
        content_length = int(self.headers['Content-Length'])
        post_data = self.rfile.read(content_length)
        data = json.loads(post_data.decode('utf-8'))

        # Get data from request
        student_name = data.get('studentName', 'Student')
        context = data.get('context', '')
        message_type = data.get('type', 'encouragement')

        # Create prompt based on type
        prompts = {
            'encouragement': f"You are {student_name}'s friendly AI study companion. {context} Give them a short, warm, encouraging message (2-3 sentences max).",
            'correction': f"You are {student_name}'s patient AI study companion. {context} Gently help them understand their mistake with a kind explanation (2-3 sentences max).",
            'celebration': f"You are {student_name}'s enthusiastic AI study companion. {context} Celebrate their success warmly (2-3 sentences max).",
            'focus': f"You are {student_name}'s supportive AI study companion. Help them refocus with a gentle reminder about staying on task (2-3 sentences max)."
        }

        prompt = prompts.get(message_type, prompts['encouragement'])

        # Call Claude API
        api_key = os.environ.get('ANTHROPIC_API_KEY')
        
        if not api_key:
            response = {
                'message': f"Great job, {student_name}! Keep going!",
                'error': 'API key not configured'
            }
            self.wfile.write(json.dumps(response).encode())
            return

        # Prepare Claude API request
        request_data = {
            'model': 'claude-sonnet-4-20250514',
            'max_tokens': 150,
            'messages': [
                {
                    'role': 'user',
                    'content': prompt
                }
            ]
        }

        # Make request to Claude API
        req = urllib.request.Request(
            'https://api.anthropic.com/v1/messages',
            data=json.dumps(request_data).encode('utf-8'),
            headers={
                'x-api-key': api_key,
                'anthropic-version': '2023-06-01',
                'content-type': 'application/json'
            }
        )

        try:
            with urllib.request.urlopen(req) as response:
                result = json.loads(response.read().decode('utf-8'))
                message = result['content'][0]['text']
                
                response_data = {
                    'message': message,
                    'studentName': student_name
                }
                
                self.wfile.write(json.dumps(response_data).encode())
        except urllib.error.HTTPError as e:
            error_response = {
                'message': f"You're doing great, {student_name}!",
                'error': f'API error: {e.code}'
            }
            self.wfile.write(json.dumps(error_response).encode())
        except Exception as e:
            error_response = {
                'message': f"Keep up the good work, {student_name}!",
                'error': str(e)
            }
            self.wfile.write(json.dumps(error_response).encode())

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

File: api/companion.js
export default async function handler(req, res) {
  // Enable CORS
  res.setHeader('Access-Control-Allow-Credentials', true);
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  if (req.method !== 'POST') {
    res.status(405).json({ error: 'Method not allowed' });
    return;
  }

  const { studentName = 'Student', context = '', type = 'encouragement' } = req.body;

  // Create prompts
  const prompts = {
    encouragement: `Give ${studentName} a short encouraging message: ${context}`,
    correction: `Gently help ${studentName} understand: ${context}`,
    celebration: `Celebrate ${studentName}'s success: ${context}`,
    focus: `Help ${studentName} stay focused: ${context}`
  };

  const prompt = prompts[type] || prompts.encouragement;

  try {
    // Call Hugging Face API
    const response = await fetch('https://api-inference.huggingface.co/models/microsoft/DialoGPT-medium', {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${process.env.HUGGINGFACE_API_KEY}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        inputs: prompt,
        parameters: {
          max_length: 100,
          temperature: 0.8
        }
      })
    });

    const result = await response.json();
    
    let message = `You're doing great, ${studentName}!`;
    
    if (Array.isArray(result) && result.length > 0) {
      message = result[0].generated_text || message;
    }

    res.status(200).json({
      message,
      studentName
    });

  } catch (error) {
    // Fallback messages
    const fallbackMessages = {
      encouragement: `You're doing great, ${studentName}! Keep trying!`,
      correction: `That's okay, ${studentName}! Let's try again together.`,
      celebration: `Amazing work, ${studentName}! You did it!`,
      focus: `Let's focus, ${studentName}. You can do this!`
    };

    res.status(200).json({
      message: fallbackMessages[type] || `Keep going, ${studentName}!`,
      error: error.message
    });
  }
}
